---
title: "sensitivity_analyses"
author: "Alish Palmos"
date: "14/04/2022"
output: html_document
editor_options:
  chunk_output_type: console
---

#SENSITIVITY ANALYSES - format GSMR output

#Set-up
```{r echo=FALSE}
rm(list=ls())
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

setwd("/scratch/groups/ukbiobank/usr/alish/AD_MR/")
```

```{r include=FALSE}

source(file = "package_check.R")

packages <- c("data.table",
              "jtools",
              "knitr",
              "gtsummary",
              "broom",
              "sandwich",
              "tidyverse",
              "ggplot2",
              "dplyr",
              "sgof",
              "TwoSampleMR",
              "qvalue",
              "kableExtra",
              "vroom",
              "openxlsx",
              "cowplot",
              "gridGraphics")

package_check(packages)

```

# Functions to get SNP infro from GSMR
```{r GSMR fucntions for creating figures, no need to take note of this}

# ************************************************** #
#              Read exposure and outcome             #
# ************************************************** #
read_gsmr_trait = function(file_con) {
    expo_str = scan(file_con, nlines=1, quiet=TRUE, what="");
    outcome_str = scan(file_con, nlines=1, quiet=TRUE, what="");
    strbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
    return(list(expo_str=expo_str, outcome_str=outcome_str))
}

# ************************************************** #
#                  Read GSMR result                  #
# ************************************************** #
read_gsmr_result = function(file_con) {
    expo_str = outcome_str = bxy = bxy_se = bxy_pval = bxy_m = c()
    while(1) {
        strbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
        if(strbuf[1] == "#gsmr_end") break;
        if(strbuf[1] == "Exposure") next;
        expo_str = c(expo_str, strbuf[1]);
        outcome_str = c(outcome_str, strbuf[2]);
        bxy = c(bxy, as.numeric(strbuf[3]));
        bxy_se = c(bxy_se, as.numeric(strbuf[4]));
        bxy_pval = c(bxy_pval, as.numeric(strbuf[5]));
        bxy_m = c(bxy_m, as.numeric(strbuf[6]));
    }
    return(cbind(expo_str, outcome_str, bxy, bxy_se, bxy_pval, bxy_m))
}

# ************************************************** #
#                  Read SNP effects                  #
# ************************************************** #
read_snp_effect = function(file_con) {
    snp_effect = c()
    while(1) {
        strbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
        if(strbuf[1] == "#effect_end") break;
        snp_effect = rbind(snp_effect, strbuf);
        print(length(strbuf))
        if(length(strbuf)<14) print(strbuf)
    }
    return(snp_effect)
}

# ************************************************** #
#                  Read SNP instruments              #
# ************************************************** #
read_snp_instru = function(file_con, snplist, nexpo, noutcome) {
    nrow = length(snplist); ncol = nexpo+noutcome
    snp_instru = matrix(NA, nrow, ncol)
    while(1) {
        strbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
        if(strbuf[1] == "#marker_end") break;
        expo_indx = as.numeric(strbuf[1]); outcome_indx = as.numeric(strbuf[2]);
        forward_flag = TRUE;
        if(expo_indx < outcome_indx) {
            outcome_indx = outcome_indx - nexpo
        } else {
            expo_indx = expo_indx - nexpo
            forward_flag = FALSE;
        }
        snpbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
        snp_indx = match(snpbuf, snplist)
        posbuf = rep(0, nrow); posbuf[snp_indx] = 1;
        indxbuf = expo_indx
        if(!forward_flag) indxbuf = indxbuf + nexpo
        if(length(which(!is.na(snp_instru[,indxbuf])))==0) {
            snp_instru[,indxbuf] = posbuf;
        } else {
            snp_instru[,indxbuf] = paste(snp_instru[,indxbuf], posbuf, sep="")
        }
    }
    return(snp_instru)
}

# ************************************************** #
#          Read output by GCTA-GSMR for plot         #
# ************************************************** #
read_gsmr_data = function(gsmr_effect_file) {
    trait_flag = gsmr_flag = marker_flag = effect_flag = FALSE;
    file_con = file(gsmr_effect_file, "r")
    while(1) {
        strbuf = scan(file_con, nlines=1, quiet=TRUE, what="");
        if(strbuf == "#trait_begin") {
            # Read the exposures and outcomes
            resbuf = read_gsmr_trait(file_con);
            expo_str = resbuf$expo_str;
            outcome_str = resbuf$outcome_str;
            pheno_str = c(expo_str, outcome_str);
            nexpo = length(expo_str); noutcome = length(outcome_str)
            trait_flag = TRUE;
        } else if(strbuf == "#gsmr_begin") {
            # Read the GSMR result
            bxy_result = read_gsmr_result(file_con);
            colnames(bxy_result) = c("Exposure", "Outcome", "bxy", "se", "p", "n_snps")
            gsmr_flag = TRUE;
        } else if(strbuf == "#effect_begin") {
            # Read the summary statistics
            snp_effect = read_snp_effect(file_con);
            snplist = as.character(snp_effect[,1])
            effect_flag = TRUE;
        } else if(strbuf == "#marker_begin") {
            # Read the SNPs
            snp_instru = read_snp_instru(file_con, snplist, nexpo, noutcome);
            snp_effect = cbind(snp_effect[,1], snp_instru, snp_effect[,-1])
            marker_flag = TRUE;
        }
        if(trait_flag==T & gsmr_flag==T & marker_flag==T & effect_flag==T) break;
    }
    return(list(pheno=c(nexpo, noutcome, pheno_str), bxy_result=bxy_result, snp_effect = snp_effect))
}

# ************************************************** #
#         Display summary of the gsmr data           #
# ************************************************** #
gsmr_summary = function(gsmr_data) {
    message("\n## Exposure and outcome")
    pheno_str = gsmr_data$pheno[c(-1,-2)]
    # exposure
    nexpo = as.numeric(gsmr_data$pheno[1]);
    noutcome = as.numeric(gsmr_data$pheno[2]);
    logger_m = paste(nexpo, "exposure(s):");
    logger_m = paste(logger_m, gsmr_data$pheno[3])
    if(nexpo > 1) {
        for(i in 2 : nexpo)
            logger_m = paste(logger_m, gsmr_data$pheno[i+2], sep=", ")
    }
    message(logger_m)
    # outcome
    logger_m = paste(noutcome, "outcome(s):");
    logger_m = paste(logger_m, gsmr_data$pheno[3+nexpo])
    if(noutcome > 1) {
        for(i in 2 : noutcome)
            logger_m = paste(logger_m, gsmr_data$pheno[i+2+nexpo], sep=", ")
    }
    message(logger_m)

    message("\n## GSMR result")
    m_bxy_rst = data.frame(gsmr_data$bxy_result)
    print(m_bxy_rst)
}


# ************************************************** #
#               Retrieve SNP effects                 #
# ************************************************** #
gsmr_snp_effect = function(gsmr_data, expo_str, outcome_str) {
   # index of SNP instruments
    pheno_str = as.character(gsmr_data$pheno[c(-1,-2)])
    nexpo = as.numeric(gsmr_data$pheno[1])
    noutcome = as.numeric(gsmr_data$pheno[2])
    expo_indx = match(expo_str, pheno_str)
    if(is.na(expo_indx)) stop("\"", expo_str, "\" is not found.")
    outcome_indx = match(outcome_str, pheno_str)
    if(is.na(outcome_indx)) stop("\"", outcome_str, "\" is not found.")
    forward_flag = TRUE;
    if(expo_indx > outcome_indx) forward_flag = FALSE;
    if(forward_flag) {
        outcome_indx = outcome_indx - nexpo;
    } else {
        expo_indx = expo_indx - nexpo;
    }
    indxbuf = expo_indx + 1
    if(!forward_flag) indxbuf = indxbuf + nexpo
    strbuf = as.character(substr(gsmr_data$snp_effect[,indxbuf], outcome_indx, outcome_indx))
    snpindx = which(strbuf=="1")
    if(length(snpindx) < 1) stop("Not enough SNPs retained.")
    # bxy
    indxbuf = which(gsmr_data$bxy_result[,1]==expo_str & gsmr_data$bxy_result[,2]==outcome_str)
    bxy = as.numeric(gsmr_data$bxy_result[indxbuf, 3])
    # SNP effects
    if(forward_flag) {
        indxbuf1 = 1 + nexpo + noutcome + 3 + (expo_indx-1)*2 + 1
        indxbuf2 = 1 + nexpo + noutcome + 3 + nexpo*2 + (outcome_indx-1)*2 + 1
    } else {
        indxbuf1 = 1 + nexpo + noutcome + 3 + nexpo*2 + (expo_indx-1)*2 + 1
        indxbuf2 = 1 + nexpo + noutcome + 3 + (outcome_indx-1)*2 + 1
    }
    snpid = as.character(gsmr_data$snp_effect[snpindx,1])
    bzx = as.numeric(gsmr_data$snp_effect[snpindx,indxbuf1]); indxbuf1 = indxbuf1 + 1;
    bzx_se = as.numeric(gsmr_data$snp_effect[snpindx,indxbuf1]);
    bzx_pval = pchisq((bzx/bzx_se)^2, 1, lower.tail=F);
    bzy = as.numeric(gsmr_data$snp_effect[snpindx,indxbuf2]); indxbuf2 = indxbuf2 + 1;
    bzy_se = as.numeric(gsmr_data$snp_effect[snpindx,indxbuf2]);
    bzy_pval = pchisq((bzy/bzy_se)^2, 1, lower.tail=F);
    return(list(snp=snpid, bxy=bxy, bzx=bzx, bzx_se=bzx_se, bzx_pval=bzx_pval, bzy=bzy, bzy_se=bzy_se, bzy_pval=bzy_pval))
}

# ************************************************** #
#                  Plot bzy vs bzx                   #
# ************************************************** #
plot_snp_effect = function(expo_str, outcome_str, bxy, bzx, bzx_se, bzy, bzy_se, effect_col=colors()[75]) {
    vals = c(bzx-bzx_se, bzx+bzx_se)
    xmin = min(vals); xmax = max(vals)
    vals = c(bzy-bzy_se, bzy+bzy_se)
    ymin = min(vals); ymax = max(vals)
    plot(bzx, bzy, pch=20, cex=0.8, bty="n", cex.axis=1.1, cex.lab=1.2,
         col=effect_col, xlim=c(xmin, xmax), ylim=c(ymin, ymax),
         xlab=substitute(paste(trait, " (", italic(b[zx]), ")", sep=""), list(trait=expo_str)),
         ylab=substitute(paste(trait, " (", italic(b[zy]), ")", sep=""), list(trait=outcome_str))
    )
    if(!is.na(bxy)) abline(0, bxy, lwd=1.5, lty=2, col="dim grey")
    ## Standard errors
    nsnps = length(bzx)
    for( i in 1:nsnps ) {
        # x axis
        xstart = bzx[i] - bzx_se[i]; xend = bzx[i] + bzx_se[i]
        ystart = bzy[i]; yend = bzy[i]
        segments(xstart, ystart, xend, yend, lwd=1.5, col=effect_col)
        # y axis
        xstart = bzx[i]; xend = bzx[i]
        ystart = bzy[i] - bzy_se[i]; yend = bzy[i] + bzy_se[i]
        segments(xstart, ystart, xend, yend, lwd=1.5, col=effect_col)


    }

    p1_recorded <- recordPlot()
ggdraw(p1_recorded)

}

# ************************************************** #
#             Plot bzy_pval vs bzx_pval              #
# ************************************************** #
plot_snp_pval = function(expo_str, outcome_str, bzx_pval, bzy_pval, gwas_thresh, truncation, effect_col) {
    eps = 1e-300; truncation = -log10(truncation);
    if(truncation > 300) {
        warning("The minimal truncated p-value would be 1e-300.")
        truncation = 300
    }
    bzx_pval = -log10(bzx_pval + eps);
    bzy_pval = -log10(bzy_pval + eps);
    pval = c(bzx_pval, bzy_pval)
    min_val = 0; max_val = max(pval);
    max_val = ifelse(max_val > truncation, truncation, max_val)
    gwas_thresh = -log10(gwas_thresh);
    plot(bzx_pval, bzy_pval, pch=20, cex=0.8, bty="n", cex.axis=1.1, cex.lab=1.2,
         col=effect_col, xlim=c(min_val, max_val), ylim=c(min_val, max_val),
         xlab=substitute(paste(trait, " (", -log[10], italic(P)[zx], ")", sep=""), list(trait=expo_str)),
         ylab=substitute(paste(trait, " (", -log[10], italic(P[zy]), ")", sep=""), list(trait=outcome_str)))
    abline(h=gwas_thresh, lty=2, lwd=1.5, col="maroon")
}

# ************************************************** #
#                Plot bxy vs bzx_pval                #
# ************************************************** #
plot_snp_bxy = function(expo_str, outcome_str, bxy, bzx_pval, effect_col) {
    eps = 1e-300;
    bzx_pval = -log10(bzx_pval + eps);
    xmin = min(bxy, na.rm=T); xmax = max(bxy, na.rm=T)
    ymin = min(bzx_pval); ymax = max(bzx_pval);
    plot(bxy, bzx_pval, pch=20, cex=0.8, bty="n", cex.axis=1.1, cex.lab=1.2,
         col=effect_col, xlim=c(xmin, xmax), ylim=c(ymin, ymax),
         xlab=substitute(paste(italic(hat(b)[xy]), " (", trait1, " -> ", trait2, ")", sep=""), list(trait1=expo_str, trait2=outcome_str)),
         ylab=substitute(paste(trait, " (", -log[10], italic(P[zx]), ")", sep=""), list(trait=expo_str)))
}

# ************************************************** #
#                  Effect size plot                  #
# ************************************************** #
# expo_str, exposure
# outcome_str, outcome
# effect_col, plotting colour
plot_gsmr_effect = function(gsmr_data, expo_str, outcome_str, effect_col=colors()[75]) {
    resbuf = gsmr_snp_effect(gsmr_data, expo_str, outcome_str);
    bxy = resbuf$bxy
    bzx = resbuf$bzx; bzx_se = resbuf$bzx_se;
    bzy = resbuf$bzy; bzy_se = resbuf$bzy_se;
    # plot
    plot_snp_effect(expo_str, outcome_str, bxy, bzx, bzx_se, bzy, bzy_se, effect_col)
}

# ************************************************** #
#                    P-value plot                    #
# ************************************************** #
# expo_str, exposure
# outcome_str, outcome
# effect_col, plotting colour
plot_gsmr_pvalue = function(gsmr_data, expo_str, outcome_str, gwas_thresh=5e-8, truncation=1e-50, effect_col=colors()[75]) {
    resbuf = gsmr_snp_effect(gsmr_data, expo_str, outcome_str);
    bzx_pval = resbuf$bzx_pval; bzy_pval = resbuf$bzy_pval;
    # plot
    plot_snp_pval(expo_str, outcome_str, bzx_pval, bzy_pval, gwas_thresh, truncation, effect_col)
}

# ************************************************** #
#                     bxy distribution plot                         #
# ************************************************** #

# expo_str, exposure
# outcome_str, outcome
# effect_col, plotting colour
plot_bxy_distribution = function(gsmr_data, expo_str, outcome_str, effect_col=colors()[75]) {
    resbuf = gsmr_snp_effect(gsmr_data, expo_str, outcome_str);
    bzx = resbuf$bzx; bzx_pval = resbuf$bzx_pval;
    bzy = resbuf$bzy;
    bxy = bzy/bzx
    # plot
    plot_snp_bxy(expo_str, outcome_str, bxy, bzx_pval, effect_col)
}

```


# PROTEIN -> AD
# Create a loop for getting all the GSMR data into a Google Sheet
```{r}

mr_df_original_AD <- vroom(file = "/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_outcome_stringent")

# mr_df_original_AD <- AD_outcome_stringent
  # arrange(aesc(qvalue))

# library(vroom)
# library(openxlsx)
# library(cowplot)

mr_df_original_AD$study <- tolower(mr_df_original_AD$study)
mr_df_original_AD$clean_names <-  gsub('\\..*', '', mr_df_original_AD$Exposure)
mr_df_original_AD$file_name <- gsub(".gsmr", "", mr_df_original_AD$file_name)

filename <- mr_df_original_AD$file_name
study <- mr_df_original_AD$study
mr_df_original_AD$names <- make.names(mr_df_original_AD$Exposure,unique=T)
names <- mr_df_original_AD$names
marker <- mr_df_original_AD$Exposure
brain <- mr_df_original_AD$Outcome
path <- "/scratch/groups/ukbiobank/usr/alish/AD_MR/"
label <- paste(mr_df_original_AD$clean_names, mr_df_original_AD$study, sep = "_")
no_dups <- make.unique(label, sep = ".")


number <- 1

rm(wb)

wb <- createWorkbook()

for(number in 1:length(filename)){

  current_marker <- filename[number]

  file_path_gsmr <- paste0(path,study[number],"_mr/","output/",current_marker,".eff_plot.gz")

  gsmr_data <- read_gsmr_data(file_path_gsmr)

  gsmr_data$pheno[3] <- label[number]
  gsmr_data$bxy_result[1] <- label[number]
  gsmr_data$bxy_result[4] <- label[number]

  graph <- plot_gsmr_effect(gsmr_data,label[number],"AD_Schwartzentruber", colors()[75])

  ggsave(filename = paste0(label[number]), plot = graph, path = paste0(path,"/sensitivity/AD_outcome/"), width = 6, height = 4, device='tiff', dpi=700)

  snp_data <- gsmr_snp_effect(gsmr_data,label[number],"AD_Schwartzentruber")

  snp_data <- as.data.frame(snp_data)

  # current_name <- substring(paste0(names[number]), 1, 30)

  addWorksheet(wb, sheetName = paste0(no_dups[number]))

  writeData(wb, sheet = paste0(no_dups[number]), snp_data, rowNames = TRUE)

}

saveWorkbook(wb, file = paste0(path,"sensitivity/protein_AD_gsmr_results.xlsx"), overwrite = TRUE)

```

#TwoSampleMR Sensitivity Analyses

```{r}

# library(vroom)
# library(openxlsx)
# library(cowplot)

mr_df_original_AD <- vroom(file = "/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_outcome_stringent")

# mr_df_original_AD <- AD_outcome_stringent
  # arrange(aesc(qvalue))

mr_df_original_AD$file_name <- gsub(".gsmr", "", mr_df_original_AD$file_name)
mr_df_original_AD$clean_names <-  gsub('\\..*', '', mr_df_original_AD$Exposure)
mr_df_original_AD$file_name <- gsub(".txt", "", mr_df_original_AD$file_name)
mr_df_original_AD$study <- tolower(mr_df_original_AD$study)

#Read in the SNPs from the excel file
filename <- mr_df_original_AD$file_name
study <- mr_df_original_AD$study
mr_df_original_AD$names <- make.names(mr_df_original_AD$Exposure,unique=T)
names <- mr_df_original_AD$names
marker <- mr_df_original_AD$Exposure
brain <- mr_df_original_AD$Outcome
path <- "/scratch/groups/ukbiobank/usr/alish/AD_MR/"
all_protein <- list.files("/scratch/groups/ukbiobank/usr/alish/AD_MR/sensitivity/AD_outcome/", include.dirs = F)
all_protein <- all_protein[-1]
label <- paste(mr_df_original_AD$clean_names, mr_df_original_AD$study, sep = "_")
filename <- toupper(filename)
no_dups <- make.unique(label, sep = ".")


# Make the columns as follows:
# SNP                 
# beta                
# SE                 
# effect_allele      
# non-effective allele
# p-value            
# effect_allele freq

number <- 1

# library(TwoSampleMR)

rm(wb)

wb <- createWorkbook()

  #using TwoSampleMR, read in the outcome of interest
  outcome_dat <- read_outcome_data(
  filename = paste0("/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_Schwartzentruber.txt"),
  sep = " ",                
  snp_col = "variant_id",          
  beta_col = "beta",         
  se_col = "standard_error",        
  effect_allele_col = "effect_allele",  
  other_allele_col = "other_allele",
  eaf = "effect_allele_frequency",
  samplesize_col = "472868",
  pval_col = "p_value")

  outcome_dat$outcome=paste0("AD")

for(number in 86:length(filename)){

  loadWorkbook(file = paste0(path,"sensitivity/protein_AD_sensitivity.xlsx"), xlsxFile = NULL, isUnzipped = FALSE)


  #read in the workbook with all our significant proteins
  current_protein <- readWorkbook(xlsxFile = paste0(path,"sensitivity/protein_AD_gsmr_results.xlsx"), sheet = paste0(label[number]), rowNames = TRUE)

  #get the SNPs of interest
  df <- current_protein %>%
    select(SNP = snp)

  #get the whole GWAS from the same protein
  original_protein <- vroom(paste0("/scratch/groups/gwas_sumstats/cleaned/blood_biomarkers/",filename[number],".gz"))

  #join the shorter and the longer protein data together
  full_df <-  inner_join(df, original_protein)

  #get all the columns that we need
  try(
    full_df <- full_df %>%
  select(SNP, BETA, SE, A1, A2, P, MAF)
  )
  try(
    full_df <- full_df %>%
  select(SNP, BETA, SE = STDERR, A1, A2, P, MAF)
  )

  #get all the columns that we need
  full_df <- full_df %>%
  select(SNP,
         beta = BETA,
         se = SE,
         effect_allele = A1,
         other_allele = A2,
         eaf = MAF
         )

  #format
  exposure_dat <- format_data(full_df, type = "exposure")

  exposure_dat$exposure=paste0(label[number])

  #Harmonisation of SNP instruments between exposures and outcomes
  harmonized <- harmonise_data(exposure_dat = exposure_dat, outcome_dat = outcome_dat, action=2)

  # get the data ready
  # all exposures -> Covid
  dataMR_keep=subset(harmonized, mr_keep==TRUE)

  # perform TwoSampleMR for all pairs
  MR_toCovid_keep<- mr(dataMR_keep, method_list = c("mr_two_sample_ml", "mr_egger_regression","mr_weighted_median", "mr_ivw", "mr_ivw_radial", "mr_ivw_mre", "mr_ivw_fe", "mr_weighted_mode"))

  #get the name of the current marker and brain volume
  current_name <- substring(paste0(no_dups[number],"_",brain[number]), 1, 30)

  #add a new sheet to the excel data frame and add the sensitivity data to this sheet
  addWorksheet(wb, sheetName = paste0(current_name))

  writeData(wb, sheet = paste0(current_name), MR_toCovid_keep, rowNames = TRUE)

  #save the results as a data frame
  Results_MR_toCovid_keep=MR_toCovid_keep # Saves the results in dataframe

  #plot the results
  plot = mr_scatter_plot(Results_MR_toCovid_keep, dataMR_keep)

  #just choose the plot from the list elements
  try({save_plot <- plot[[1]]})

  #save plot
  try({ggsave(filename = paste0(no_dups[number],"_",brain[number]), plot = save_plot, path = paste0(path,"sensitivity/AD_outcome/2SMR/plots/"), width = 6, height = 4, device='tiff', dpi=700)})

}

#save the entire excel workbook with all the sensitivity analyses
saveWorkbook(wb, file = paste0(path,"sensitivity/protein_AD_sensitivity.xlsx"), overwrite = TRUE)

```

# AD -> PROTEIN
# Create a loop for getting all the GSMR data into a Google Sheet
```{r}

mr_df_original_AD <- vroom("/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_exposure_stringent")

# mr_df_original_AD <- AD_exposure_stringent
  # arrange(aesc(qvalue))

# library(vroom)
# library(openxlsx)
# library(cowplot)

mr_df_original_AD$study <- tolower(mr_df_original_AD$study)
mr_df_original_AD$clean_names <-  gsub('\\..*', '', mr_df_original_AD$Outcome)
# mr_df_original_AD$marker_name <- sub('\\_.*', '', mr_df_original_AD$marker_name)
mr_df_original_AD$file_name <- gsub(".gsmr", "", mr_df_original_AD$file_name)

filename <- mr_df_original_AD$file_name
study <- mr_df_original_AD$study
mr_df_original_AD$names <- make.names(mr_df_original_AD$Outcome,unique=T)
names <- mr_df_original_AD$names
marker <- mr_df_original_AD$Outcome
brain <- mr_df_original_AD$Exposure
path <- "/scratch/groups/ukbiobank/usr/alish/AD_MR/"
label <- paste(mr_df_original_AD$clean_names, mr_df_original_AD$study, sep = "_")
no_dups <- make.unique(label, sep = ".")

number <- 1

rm(wb)

wb <- createWorkbook()

for(number in 1:length(filename)){

  current_marker <- filename[number]

  print(filename[number])

  file_path_gsmr <- paste0(path,study[number],"_mr/","/ADexp/output/",current_marker,".eff_plot.gz")

  gsmr_data <- read_gsmr_data(file_path_gsmr)

  gsmr_data$pheno[3] <- label[number]
  gsmr_data$bxy_result[1] <- label[number]
  gsmr_data$bxy_result[4] <- label[number]

  graph <- plot_gsmr_effect(gsmr_data,"AD_Schwartzentruber",label[number], colors()[75])

  ggsave(filename = paste0(label[number]), plot = graph, path = paste0(path,"/sensitivity/AD_exposure/"), width = 6, height = 4, device='tiff', dpi=700)

  snp_data <- gsmr_snp_effect(gsmr_data,"AD_Schwartzentruber",label[number])

  snp_data <- as.data.frame(snp_data)

  addWorksheet(wb, sheetName = paste0(no_dups[number]))

  writeData(wb, sheet = paste0(no_dups[number]), snp_data, rowNames = TRUE)


}

saveWorkbook(wb, file = paste0(path,"sensitivity/AD_protein_gsmr_results.xlsx"), overwrite = TRUE)

```

#TwoSampleMR Sensitivity Analyses
```{r}

mr_df_original_AD <- vroom("/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_exposure_stringent")

# mr_df_original_AD <- AD_exposure_stringent
  # arrange(aesc(qvalue))

mr_df_original_AD$file_name <- gsub(".gsmr", "", mr_df_original_AD$file_name)
mr_df_original_AD$clean_names <-  gsub('\\..*', '', mr_df_original_AD$Outcome)
mr_df_original_AD$file_name <- gsub(".txt", "", mr_df_original_AD$file_name)
mr_df_original_AD$Outcome <- sub('\\..*', '', mr_df_original_AD$Outcome)
mr_df_original_AD$name <- paste(mr_df_original_AD$Outcome, mr_df_original_AD$study, sep = "_")
mr_df_original_AD$study <- tolower(mr_df_original_AD$study)

#Read in the SNPs from the excel file
filename <- mr_df_original_AD$file_name
study <- mr_df_original_AD$study
mr_df_original_AD$names <- make.names(mr_df_original_AD$Outcome,unique=T)
names <- mr_df_original_AD$names
marker <- mr_df_original_AD$Outcome
brain <- mr_df_original_AD$Exposure
path <- "/scratch/groups/ukbiobank/usr/alish/AD_MR/"
all_protein <- list.files("/scratch/groups/ukbiobank/usr/alish/AD_MR/sensitivity/AD_exposure/", include.dirs = F)
all_protein <- all_protein[-1]
label <- paste(mr_df_original_AD$clean_names, mr_df_original_AD$study, sep = "_")
filename <- toupper(filename)
no_dups <- make.unique(label, sep = ".")


#this is only done once to get the smaller AD GWAS
# AD_gwas_sensitivity <- vroom("/scratch/groups/ukbiobank/usr/alish/AD_MR/AD_Schwartzentruber.txt")
#
# AD_gwas_sensitivity <- AD_gwas_sensitivity %>%
#   subset(p_value<5*10^-8)
# write.table(AD_gwas_sensitivity, file = paste0("/scratch/groups/ukbiobank/usr/alish/AD_MR/sensitivity/AD_gwas_sensitivity.txt"), col.names=T, row.names=F, quote=F,sep = "\t")


# Make the columns as follows:
# SNP                 
# beta                
# SE                 
# effect_allele      
# non-effective allele
# p-value            
# effect_allele freq

number <- 1

# library(TwoSampleMR)

rm(wb)

wb <- createWorkbook()

for(number in 1:length(filename)){

  #using TwoSampleMR, read in the outcome of interest
  exposure_dat <- read_exposure_data(
  filename = paste0("/scratch/groups/ukbiobank/usr/alish/AD_MR/sensitivity/AD_gwas_sensitivity.txt"),
  sep = "\t",                
  snp_col = "variant_id",          
  beta_col = "beta",         
  se_col = "standard_error",        
  effect_allele_col = "effect_allele",  
  other_allele_col = "other_allele",
  eaf = "effect_allele_frequency",
  samplesize_col = "472868",
  pval_col = "p_value")

  exposure_dat$exposure=paste0("AD")

  #read in the workbook with all our significant proteins
  current_protein <- readWorkbook(xlsxFile = paste0(path,"sensitivity/protein_AD_sensitivity.xlsx"), sheet = paste0(marker[number]), rowNames = TRUE)

  #get the whole GWAS from the same protein
  original_protein <- vroom(paste0("/scratch/groups/gwas_sumstats/cleaned/blood_biomarkers/",filename[number],".gz"))

  #get all the columns that we need
  full_df <- original_protein %>%
  select(SNP,
         beta = BETA,
         se = SE,
         effect_allele = A1,
         other_allele = A2,
         eaf = MAF
         )

  #format
  outcome_dat <- format_data(full_df, type = "outcome")

  outcome_dat$outcome=paste0(marker[number])

  exposure_dat <- clump_data(exposure_dat, clump_r2 = 0.05)

  #Harmonisation of SNP instruments between exposures and outcomes
  harmonized <- harmonise_data(exposure_dat = exposure_dat, outcome_dat = outcome_dat, action=2)

  # get the data ready
  # all exposures -> Covid
  dataMR_keep=subset(harmonized, mr_keep==TRUE)

  # perform TwoSampleMR for all pairs
  MR_toCovid_keep<- mr(dataMR_keep)

  #get the name of the current marker and brain volume
  current_name <- substring(paste0(brain[number],"_",label[number]), 1, 30)

  #add a new sheet to the excel data frame and add the sensitivity data to this sheet
  addWorksheet(wb, sheetName = paste0(current_name))

  writeData(wb, sheet = paste0(current_name), MR_toCovid_keep, rowNames = TRUE)

  #save the results as a data frame
  Results_MR_toCovid_keep=MR_toCovid_keep # Saves the results in dataframe

  #plot the results
  plot = mr_scatter_plot(Results_MR_toCovid_keep, dataMR_keep)

  #just choose the plot from the list elements
  try({save_plot <- plot[[1]]})

  #save plot
  try({ggsave(filename = paste0(brain[number],"_",label[number]), plot = save_plot, path = paste0(path,"sensitivity/AD_exposure/2SMR/plots/"), width = 6, height = 4, device='tiff', dpi=700)})

}

#save the entire excel workbook with all the sensitivity analyses
saveWorkbook(wb, file = paste0(path,"sensitivity/AD_protein_brain_sensitivity.xlsx"), overwrite = TRUE)

```
